### PR TITLE
prospective-parachains: introduce `backed_in_path_only` flag for advertisements

### PR DESCRIPTION
**Pull request by slumber**
Closes https://github.com/paritytech/polkadot/issues/5923

### CODE REVIEW

File: node/subsystem-types/src/messages.rs [hunk 1]
```diff
@@ -963,6 +963,9 @@ pub struct HypotheticalFrontierRequest {
 	pub candidates: Vec<HypotheticalCandidate>,
 	/// Either a specific fragment tree to check, otherwise all.
 	pub fragment_tree_relay_parent: Option<Hash>,
+	/// Only return membership if all candidates in the path from the
+	/// root are backed.
+	pub backed_in_path_only: bool,
```
Comment (slumber) [inline node/subsystem-types/src/messages.rs:968]: @rphmeier I'm out of context with statement distribution, should it be a default option?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2333141

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2333798

File: node/core/prospective-parachains/src/fragment_tree.rs [hunk 2]
```diff
@@ -518,10 +540,15 @@ impl FragmentTree {
 		&self,
 		hash: CandidateHash,
 		candidate: HypotheticalCandidate,
+		candidate_storage: &CandidateStorage,
+		backed_in_path_only: bool,
```
Comment (mrcnski) [inline node/core/prospective-parachains/src/fragment_tree.rs:544]: nit: Doc can be updated to reflect this new option.

**Comment by slumber** [comment]
> Out of curiosity: if we have a path `root -> a -> b`, then is it possible for `a` to be not backed if `b` is backed? How can that happen?

When we accept statements from other validators we don't require that parent candidates are backed, but only that their membership is recognised. So probably it can happen due to some network lag.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2334837

File: node/subsystem-types/src/messages.rs (hunk 1 above)
Comment (rphmeier) [reply node/subsystem-types/src/messages.rs:968]: What do you mean by a default option?

Comment (slumber) [reply node/subsystem-types/src/messages.rs:968]: Do we only want to set it to `true` when checking incoming collator's advertisements, or is it always the case?

**Review by rphmeier — APPROVED** [review]

**Pull request merged**